File: app/src/main/java/xyz/rodit/snapmod/features/chatmenu/ChatMenuModifier.java
package xyz.rodit.snapmod.features.chatmenu;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import de.robv.android.xposed.XC_MethodHook;
import xyz.rodit.snapmod.features.Feature;
import xyz.rodit.snapmod.features.FeatureContext;
import xyz.rodit.snapmod.mappings.FriendChatActionHandler;
import xyz.rodit.snapmod.mappings.FriendChatActionMenuBuilder;
import xyz.rodit.snapmod.mappings.RxSingleton;
import xyz.rodit.snapmod.mappings.SendChatAction;

public class ChatMenuModifier extends Feature {

    protected static final String EVENT_PREFIX = "CUSTOM_ACTION";
    protected static final String EVENT_DELIMITER = "\0:\0";

    private final Map<String, MenuPlugin> plugins = new HashMap<>();

    public ChatMenuModifier(FeatureContext context) {
        super(context);
    }

    @Override
    protected void init() {
        registerPlugin(new PinOption(context));
        registerPlugin(new StealthOption(context));
    }

    private void registerPlugin(MenuPlugin plugin) {
        plugins.put(plugin.getName(), plugin);
    }

    @Override
    protected void performHooks() {
        // Add plugin actions to chat action menu.
        FriendChatActionMenuBuilder.build.hook(new XC_MethodHook() {
            @Override
            protected void afterHookedMethod(MethodHookParam param) {
                FriendChatActionMenuBuilder $this = FriendChatActionMenuBuilder.wrap(param.thisObject);
                String key = $this.getFeedInfoHolder().getInfo().getKey();
                List options = (List) RxSingleton.wrap(param.getResult()).getValue();

                for (MenuPlugin plugin : plugins.values()) {
                    if (plugin.shouldCreate()) {
                        options.add(plugin.createModel(key));
                    }
                }
            }
        });

        // Override plugin action events.
        FriendChatActionHandler.handle.hook(new XC_MethodHook() {
            @Override
            protected void beforeHookedMethod(MethodHookParam param) {
                if (!SendChatAction.isInstance(param.args[0])) {
                    return;
                }

                SendChatAction action = SendChatAction.wrap(param.args[0]);
                String data = action.getDataModel().getKey();
                if (data.startsWith(EVENT_PREFIX)) {
                    String[] parts = data.split(EVENT_DELIMITER);
                    if (parts[0].equals(EVENT_PREFIX) && parts.length == 3) {
                        String pluginName = parts[1];
                        if (!plugins.containsKey(pluginName)) {
                            return;
                        }

                        MenuPlugin plugin = plugins.get(pluginName);
                        plugin.handleEvent(parts[2]);
                        param.setResult(null);
                    }
                }
            }
        });
    }
}

File: app/src/main/java/xyz/rodit/snapmod/features/chatmenu/ChatMenuModifier.kt
package xyz.rodit.snapmod.features.chatmenu

import xyz.rodit.snapmod.CustomResources.string.menu_option_auto_download
import xyz.rodit.snapmod.CustomResources.string.menu_option_auto_save
import xyz.rodit.snapmod.CustomResources.string.menu_option_stealth_mode
import xyz.rodit.snapmod.Shared
import xyz.rodit.snapmod.features.Feature
import xyz.rodit.snapmod.features.FeatureContext
import xyz.rodit.snapmod.mappings.FriendChatActionHandler
import xyz.rodit.snapmod.mappings.FriendChatActionMenuBuilder
import xyz.rodit.snapmod.mappings.RxSingleton
import xyz.rodit.snapmod.mappings.SendChatAction
import xyz.rodit.snapmod.util.after
import xyz.rodit.snapmod.util.before

const val EVENT_PREFIX = "CUSTOM_ACTION"
const val EVENT_DELIMITER = "\u0000:\u0000"
const val PIN_STRING_NAME = "action_menu_pin_conversation"

class ChatMenuModifier(context: FeatureContext) : Feature(context) {

    private val plugins = mutableMapOf<String, MenuPlugin>()

    override fun init() {
        registerPlugin(PreviewOption(context))
        registerPlugin(ExportOption(context))

        val pinTextResource = context.appContext.resources.getIdentifier(
            PIN_STRING_NAME,
            "string",
            Shared.SNAPCHAT_PACKAGE
        )
        registerConversationToggle("pinning", pinTextResource) { it.pinned }
        registerConversationToggle("stealth", menu_option_stealth_mode) { it.stealth }
        registerConversationToggle("auto_save", menu_option_auto_save) { it.autoSave }
        registerConversationToggle("auto_download", menu_option_auto_download) { it.autoDownload }
    }

    private fun registerConversationToggle(name: String, textResource: Int, manager: Manager) {
        registerPlugin(ConversationToggleOption(context, name, textResource, manager))
    }

    private fun registerPlugin(plugin: MenuPlugin) {
        plugins[plugin.name] = plugin
    }

    override fun performHooks() {
        // Add plugin actions to chat action menu.
        FriendChatActionMenuBuilder.build.after {
            val self = FriendChatActionMenuBuilder.wrap(it.thisObject)
            val key = self.feedInfoHolder.info.key
            val options = RxSingleton.wrap(it.result).value as MutableList<Any>

            options.addAll(
                plugins.values.filter(MenuPlugin::shouldCreate).map { p -> p.createModel(key) })
        }

        // Override plugin action events.
        FriendChatActionHandler.handle.before {
            if (!SendChatAction.isInstance(it.args[0])) return@before

            val action = SendChatAction.wrap(it.args[0])
            val data = action.dataModel.key
            val parts = data.split(EVENT_DELIMITER)
            if (parts[0] == EVENT_PREFIX && parts.size == 3) {
                val pluginName = parts[1]
                plugins[pluginName]?.let { p ->
                    p.handleEvent(parts[2])
                    it.result = null
                }
            }
        }

        plugins.values.forEach(MenuPlugin::performHooks)
    }
}